File: jas/tools/anchor_point_tool.py
"""Anchor Point (Convert) tool.

Three interactions, mirroring jas_dioxus/src/tools/anchor_point_tool.rs:

- Drag on a corner anchor: pull out symmetric handles -> smooth.
- Click on a smooth anchor: collapse handles to anchor -> corner.
- Drag on a control handle: move it independently -> cusp.

Hit-test priority: handles before anchors.
"""

from __future__ import annotations

import dataclasses
import math
from typing import TYPE_CHECKING

from geometry.element import (
    ClosePath, CurveTo, LineTo, MoveTo, Path, PathCommand,
    convert_corner_to_smooth, convert_smooth_to_corner,
    is_smooth_point, move_path_handle_independent, path_handle_positions,
)
from document.document import ElementSelection
from tools.tool import CanvasTool, HIT_RADIUS

if TYPE_CHECKING:
    from PySide6.QtGui import QPainter
    from tools.tool import ToolContext


def _anchor_points(d: tuple[PathCommand, ...]) -> list[tuple[float, float]]:
    """Anchor positions of a path, indexed by anchor index (closePath skipped)."""
    pts: list[tuple[float, float]] = []
    for cmd in d:
        if isinstance(cmd, ClosePath):
            continue
        if isinstance(cmd, (MoveTo, LineTo)):
            pts.append((cmd.x, cmd.y))
        elif isinstance(cmd, CurveTo):
            pts.append((cmd.x, cmd.y))
        else:
            # Other path command types: use endpoint via attribute access.
            x = getattr(cmd, "x", None)
            y = getattr(cmd, "y", None)
            if x is not None and y is not None:
                pts.append((x, y))
    return pts


def _find_anchor_idx_at(
    d: tuple[PathCommand, ...], px: float, py: float,
) -> int | None:
    """Find the anchor index of a point near (px, py)."""
    for i, (ax, ay) in enumerate(_anchor_points(d)):
        if math.hypot(px - ax, py - ay) < HIT_RADIUS:
            return i
    return None


def _find_handle_at(
    d: tuple[PathCommand, ...], px: float, py: float,
) -> tuple[int, str, float, float] | None:
    """Find a handle (anchor_idx, handle_type, hx, hy) near (px, py)."""
    n = len(_anchor_points(d))
    for ai in range(n):
        h_in, h_out = path_handle_positions(d, ai)
        if h_in is not None:
            hx, hy = h_in
            if math.hypot(px - hx, py - hy) < HIT_RADIUS:
                return (ai, "in", hx, hy)
        if h_out is not None:
            hx, hy = h_out
            if math.hypot(px - hx, py - hy) < HIT_RADIUS:
                return (ai, "out", hx, hy)
    return None


def _each_path(ctx: ToolContext):
    """Yield (path_tuple, element) for every Path element, walking one
    level into unlocked groups."""
    from geometry.element import Group
    doc = ctx.model.document
    for li, layer in enumerate(doc.layers):
        for ci, child in enumerate(layer.children):
            if isinstance(child, Path):
                yield (li, ci), child
            elif isinstance(child, Group) and not child.locked:
                for gi, gc in enumerate(child.children):
                    if isinstance(gc, Path):
                        yield (li, ci, gi), gc


def _hit_test_anchor(
    ctx: ToolContext, x: float, y: float,
) -> tuple[tuple[int, ...], Path, int] | None:
    for path, pe in _each_path(ctx):
        idx = _find_anchor_idx_at(pe.d, x, y)
        if idx is not None:
            return path, pe, idx
    return None


def _hit_test_handle(
    ctx: ToolContext, x: float, y: float,
) -> tuple[tuple[int, ...], Path, int, str, float, float] | None:
    for path, pe in _each_path(ctx):
        hit = _find_handle_at(pe.d, x, y)
        if hit is not None:
            ai, ht, hx, hy = hit
            return path, pe, ai, ht, hx, hy
    return None


def _select_all_cps(ctx: ToolContext, path: tuple[int, ...]) -> None:
    doc = ctx.model.document
    new_sel_entry = ElementSelection.all(path)
    new_selection = frozenset(es for es in doc.selection if es.path != path) | {new_sel_entry}
    ctx.model.document = dataclasses.replace(doc, selection=new_selection)


def _replace(ctx: ToolContext, path: tuple[int, ...], pe: Path, new_pe: Path) -> None:
    doc = ctx.model.document.replace_element(path, new_pe)
    ctx.model.document = doc
    ctx.request_update()


class AnchorPointTool(CanvasTool):
    """Convert anchor points between smooth, corner, and cusp states."""

    def __init__(self) -> None:
        # State is one of:
        #   None
        #   ("dragging_corner", path, pe, anchor_idx, sx, sy)
        #   ("dragging_handle", path, pe, anchor_idx, handle_type, shx, shy)
        #   ("pressed_smooth", path, pe, anchor_idx, sx, sy)
        self._state: tuple | None = None

    def on_press(
        self, ctx: ToolContext, x: float, y: float,
        shift: bool = False, alt: bool = False,
    ) -> None:
        # Handle hit takes priority (cusp behaviour).
        h = _hit_test_handle(ctx, x, y)
        if h is not None:
            path, pe, ai, ht, hx, hy = h
            self._state = ("dragging_handle", path, pe, ai, ht, hx, hy)
            return
        a = _hit_test_anchor(ctx, x, y)
        if a is None:
            return
        path, pe, ai = a
        if is_smooth_point(pe.d, ai):
            self._state = ("pressed_smooth", path, pe, ai, x, y)
        else:
            self._state = ("dragging_corner", path, pe, ai, x, y)

    def on_move(
        self, ctx: ToolContext, x: float, y: float,
        shift: bool = False, dragging: bool = False,
    ) -> None:
        if not dragging or self._state is None:
            return
        kind = self._state[0]
        if kind == "dragging_corner":
            _, path, pe, ai, _sx, _sy = self._state
            new_pe = convert_corner_to_smooth(pe, ai, x, y)
            _replace(ctx, path, pe, new_pe)
        elif kind == "dragging_handle":
            _, path, pe, ai, ht, shx, shy = self._state
            dx = x - shx
            dy = y - shy
            new_pe = move_path_handle_independent(pe, ai, ht, dx, dy)
            _replace(ctx, path, pe, new_pe)
        elif kind == "pressed_smooth":
            _, path, pe, ai, sx, sy = self._state
            if math.hypot(x - sx, y - sy) > 3.0:
                corner_pe = convert_smooth_to_corner(pe, ai)
                new_pe = convert_corner_to_smooth(corner_pe, ai, x, y)
                _replace(ctx, path, pe, new_pe)
                self._state = ("dragging_corner", path, corner_pe, ai, sx, sy)

    def on_release(
        self, ctx: ToolContext, x: float, y: float,
        shift: bool = False, alt: bool = False,
    ) -> None:
        s = self._state
        self._state = None
        if s is None:
            return
        kind = s[0]
        if kind == "pressed_smooth":
            _, path, pe, ai, _sx, _sy = s
            ctx.snapshot()
            new_pe = convert_smooth_to_corner(pe, ai)
            _replace(ctx, path, pe, new_pe)
            _select_all_cps(ctx, path)
        elif kind == "dragging_corner":
            _, path, pe, ai, sx, sy = s
            if math.hypot(x - sx, y - sy) > 1.0:
                ctx.snapshot()
                new_pe = convert_corner_to_smooth(pe, ai, x, y)
                _replace(ctx, path, pe, new_pe)
                _select_all_cps(ctx, path)
        elif kind == "dragging_handle":
            _, path, pe, ai, ht, shx, shy = s
            dx = x - shx
            dy = y - shy
            if abs(dx) > 0.5 or abs(dy) > 0.5:
                ctx.snapshot()
                new_pe = move_path_handle_independent(pe, ai, ht, dx, dy)
                _replace(ctx, path, pe, new_pe)
                _select_all_cps(ctx, path)

    def draw_overlay(self, ctx: ToolContext, painter: "QPainter") -> None:
        pass

File: JasSwift/Sources/Tools/AnchorPointTool.swift
import AppKit
import Foundation

// MARK: - Anchor Point (Convert) Tool
//
// Three interactions, mirroring jas_dioxus/src/tools/anchor_point_tool.rs:
//
// - Drag on a corner anchor: pull out symmetric control handles → smooth.
// - Click on a smooth anchor (no drag): collapse handles to anchor → corner.
// - Drag on a control handle: move that handle independently → cusp.
//
// Hit-test priority: handles before anchors.

private enum AnchorPointState {
    case idle
    /// Dragging from a corner anchor to pull out handles.
    case draggingCorner(path: ElementPath, pe: Path, anchorIdx: Int, startX: Double, startY: Double)
    /// Dragging a control handle independently (cusp).
    case draggingHandle(path: ElementPath, pe: Path, anchorIdx: Int, handleType: String,
                        startHx: Double, startHy: Double)
    /// Pressed on a smooth anchor; converts to corner on release if no drag.
    case pressedSmooth(path: ElementPath, pe: Path, anchorIdx: Int, startX: Double, startY: Double)
}

final class AnchorPointTool: CanvasTool {
    private var state: AnchorPointState = .idle

    // MARK: - Hit testing

    /// Iterate every Path element in the document, walking into unlocked
    /// groups one level deep, and return the first one for which `body`
    /// produces a non-nil result. Returns `(elementPath, hit)`.
    private static func eachPath<R>(
        _ doc: Document,
        _ body: (Path) -> R?
    ) -> (ElementPath, Path, R)? {
        for (li, layer) in doc.layers.enumerated() {
            for (ci, child) in layer.children.enumerated() {
                if case .path(let pe) = child, let r = body(pe) {
                    return ([li, ci], pe, r)
                }
                if case .group(let g) = child, !g.locked {
                    for (gi, gc) in g.children.enumerated() {
                        if case .path(let pe) = gc, let r = body(pe) {
                            return ([li, ci, gi], pe, r)
                        }
                    }
                }
            }
        }
        return nil
    }

    /// Anchor positions of a path, indexed by anchor index (skipping closePath).
    private static func anchorPoints(_ d: [PathCommand]) -> [(Double, Double)] {
        var pts: [(Double, Double)] = []
        for cmd in d {
            switch cmd {
            case .moveTo(let x, let y), .lineTo(let x, let y):
                pts.append((x, y))
            case .curveTo(_, _, _, _, let x, let y):
                pts.append((x, y))
            case .closePath:
                continue
            default:
                if let ep = cmd.endpoint { pts.append(ep) }
            }
        }
        return pts
    }

    /// Find an anchor index whose position is within `hitRadius` of (px, py).
    private static func findAnchorAt(_ pe: Path, px: Double, py: Double) -> Int? {
        let pts = anchorPoints(pe.d)
        for (i, p) in pts.enumerated() {
            if hypot(px - p.0, py - p.1) < Double(hitRadius) {
                return i
            }
        }
        return nil
    }

    /// Hit-test handles on a single path. Returns (anchorIdx, handleType, hx, hy).
    private static func findHandleAt(_ pe: Path, px: Double, py: Double)
        -> (Int, String, Double, Double)? {
        let anchorCount = anchorPoints(pe.d).count
        for ai in 0..<anchorCount {
            let (hIn, hOut) = pathHandlePositions(pe.d, anchorIdx: ai)
            if let (hx, hy) = hIn,
               hypot(px - hx, py - hy) < Double(hitRadius) {
                return (ai, "in", hx, hy)
            }
            if let (hx, hy) = hOut,
               hypot(px - hx, py - hy) < Double(hitRadius) {
                return (ai, "out", hx, hy)
            }
        }
        return nil
    }

    // MARK: - Selection helper

    private static func selectAllCps(_ ctx: ToolContext, path: ElementPath) {
        let doc = ctx.document
        var newSelection = doc.selection.filter { $0.path != path }
        newSelection.insert(ElementSelection.all(path))
        let newDoc = Document(layers: doc.layers,
                              selectedLayer: doc.selectedLayer,
                              selection: newSelection)
        ctx.controller.setDocument(newDoc)
    }

    /// Apply a path-command transformation to the path at `path` and push
    /// the result back into the model. Preserves all non-`d` Path fields.
    private static func replacePathCommands(_ ctx: ToolContext,
                                            path: ElementPath,
                                            pe: Path,
                                            newCmds: [PathCommand]) {
        let newElem = Element.path(Path(d: newCmds, fill: pe.fill, stroke: pe.stroke,
                                        widthPoints: pe.widthPoints,
                                        opacity: pe.opacity, transform: pe.transform,
                                        locked: pe.locked, visibility: pe.visibility))
        let newDoc = ctx.document.replaceElement(path, with: newElem)
        ctx.controller.setDocument(newDoc)
        ctx.requestUpdate()
    }

    // MARK: - CanvasTool protocol

    func onPress(_ ctx: ToolContext, x: Double, y: Double, shift: Bool, alt: Bool) {
        let doc = ctx.document

        // 1. Handle hit takes priority (cusp behaviour).
        if let (path, pe, hit) = Self.eachPath(doc, { Self.findHandleAt($0, px: x, py: y) }) {
            let (ai, handleType, hx, hy) = hit
            state = .draggingHandle(path: path, pe: pe, anchorIdx: ai,
                                    handleType: handleType, startHx: hx, startHy: hy)
            return
        }

        // 2. Anchor hit.
        if let (path, pe, ai) = Self.eachPath(doc, { Self.findAnchorAt($0, px: x, py: y) }) {
            if isSmoothPoint(pe.d, anchorIdx: ai) {
                state = .pressedSmooth(path: path, pe: pe, anchorIdx: ai, startX: x, startY: y)
            } else {
                state = .draggingCorner(path: path, pe: pe, anchorIdx: ai, startX: x, startY: y)
            }
        }
    }

    func onMove(_ ctx: ToolContext, x: Double, y: Double, shift: Bool, dragging: Bool) {
        guard dragging else { return }
        switch state {
        case .draggingCorner(let path, let pe, let ai, _, _):
            // Live preview: pull out symmetric handles toward (x, y).
            let newCmds = convertCornerToSmooth(pe.d, anchorIdx: ai, hx: x, hy: y)
            Self.replacePathCommands(ctx, path: path, pe: pe, newCmds: newCmds)

        case .draggingHandle(let path, let pe, let ai, let handleType, let startHx, let startHy):
            let dx = x - startHx
            let dy = y - startHy
            let newCmds = movePathHandleIndependent(pe.d, anchorIdx: ai,
                                                    handleType: handleType, dx: dx, dy: dy)
            Self.replacePathCommands(ctx, path: path, pe: pe, newCmds: newCmds)

        case .pressedSmooth(let path, let pe, let ai, let sx, let sy):
            // Once the user has dragged > 3px from a smooth point, behave
            // like a corner-drag: collapse handles to anchor first, then
            // pull out new handles toward the cursor.
            let dist = hypot(x - sx, y - sy)
            if dist > 3.0 {
                let cornerCmds = convertSmoothToCorner(pe.d, anchorIdx: ai)
                let newCmds = convertCornerToSmooth(cornerCmds, anchorIdx: ai, hx: x, hy: y)
                Self.replacePathCommands(ctx, path: path, pe: pe, newCmds: newCmds)
                let cornerPe = Path(d: cornerCmds, fill: pe.fill, stroke: pe.stroke,
                                    opacity: pe.opacity, transform: pe.transform,
                                    locked: pe.locked, visibility: pe.visibility)
                state = .draggingCorner(path: path, pe: cornerPe, anchorIdx: ai,
                                        startX: sx, startY: sy)
            }

        case .idle:
            break
        }
    }

    func onRelease(_ ctx: ToolContext, x: Double, y: Double, shift: Bool, alt: Bool) {
        let s = state
        state = .idle
        switch s {
        case .pressedSmooth(let path, let pe, let ai, _, _):
            // Click on a smooth point → convert to corner.
            ctx.snapshot()
            let newCmds = convertSmoothToCorner(pe.d, anchorIdx: ai)
            Self.replacePathCommands(ctx, path: path, pe: pe, newCmds: newCmds)
            Self.selectAllCps(ctx, path: path)

        case .draggingCorner(let path, let pe, let ai, let sx, let sy):
            let dist = hypot(x - sx, y - sy)
            if dist > 1.0 {
                ctx.snapshot()
                let newCmds = convertCornerToSmooth(pe.d, anchorIdx: ai, hx: x, hy: y)
                Self.replacePathCommands(ctx, path: path, pe: pe, newCmds: newCmds)
                Self.selectAllCps(ctx, path: path)
            }

        case .draggingHandle(let path, let pe, let ai, let handleType, let startHx, let startHy):
            let dx = x - startHx
            let dy = y - startHy
            if abs(dx) > 0.5 || abs(dy) > 0.5 {
                ctx.snapshot()
                let newCmds = movePathHandleIndependent(pe.d, anchorIdx: ai,
                                                        handleType: handleType, dx: dx, dy: dy)
                Self.replacePathCommands(ctx, path: path, pe: pe, newCmds: newCmds)
                Self.selectAllCps(ctx, path: path)
            }

        case .idle:
            break
        }
    }

    func drawOverlay(_ ctx: ToolContext, _ cgCtx: CGContext) {
        // No overlay; the canvas already draws path handles for the active path.
    }
}
